### PR TITLE
Add 4 basic DSC test cases.

### PR DESCRIPTION
**Pull request by v-yanbgu**
@shmao , @huanwu  Please help to review this pull request.

### CODE REVIEW

File: src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj [hunk 1]
```diff
@@ -31,10 +31,14 @@
...
       <Link>CommonTest\System\PlatformDetection.cs</Link>
     </Compile>
+    <Compile Include="SerializationTestTypes\ObjRefSample.cs" />
   </ItemGroup>
   <ItemGroup Condition="'$(TargetGroup)' != 'uap'">
     <Compile Include="$(TestSourceFolder)SerializationTypes.CoreCLR.cs" />
     <Compile Include="$(TestSourceFolder)DataContractSerializer.CoreCLR.cs" />
   </ItemGroup>
+  <ItemGroup>
+    <Reference Include="System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089, processorArchitecture=MSIL" />
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj]: Why do we need to reference this?

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/Primitives.cs [hunk 2]
```diff
@@ -80,4 +80,24 @@ public class EnumStructContainer
...
+            age = 6;
+            name = "smith";
+        }
+
+        public Person()
+        {
+        }
+
+        [DataMember]
+        public int age;
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/Primitives.cs]: age -> Age

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/Primitives.cs [hunk 3]
```diff
@@ -80,4 +80,24 @@ public class EnumStructContainer
...
+
+        [DataMember]
+        public string name;
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/Primitives.cs]: name -> Name

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs [hunk 4]
```diff
@@ -86,4 +86,40 @@ public override bool TryResolveType(Type dataContractType, Type declaredType, Da
...
+        public override Type ResolveName(string typeName, string typeNamespace, Type declaredType, DataContractResolver KTResolver)
+        {
+            throw new NotImplementedException("Deserialization is supposed to be handled by the SimpleResolver_DeSer resolver.");
+        }
+    }
+
+    [Serializable]
+    public class SimpleResolver_DeSer : DataContractResolver
+    {
+        public string defaultNS = "http://schemas.datacontract.org/2004/07/";
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs]: defaultNS -> DefaultNS (public field)

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs [hunk 5]
```diff
@@ -86,4 +86,40 @@ public override bool TryResolveType(Type dataContractType, Type declaredType, Da
             return knownTypeResolver.TryResolveType(actualDataContractType, declaredType, null, out typeName, out typeNamespace);
         }
     }
+
+    [Serializable]
+    public class SimpleResolver_Ser : DataContractResolver
+    {
+        public string defaultNS = "http://schemas.datacontract.org/2004/07/";
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs]: defaultNS -> DefaultNS (public field)

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 6]
```diff
@@ -2860,6 +2860,88 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
...
+        dcrVariationsGoing.unknownType1 = new SerializationTestTypes.Person();
+        dcrVariationsGoing.unknownType2 = new SerializationTestTypes.SimpleDC();
+        var setting1 = new DataContractSerializerSettings() { DataContractResolver = new SerializationTestTypes.SimpleResolver_Ser(), MaxItemsInObjectGraph = int.MaxValue, IgnoreExtensionDataObject = false, PreserveObjectReferences = true};
+        var setting2 = new DataContractSerializerSettings() { DataContractResolver = new SerializationTestTypes.SimpleResolver_DeSer(), MaxItemsInObjectGraph = int.MaxValue, IgnoreExtensionDataObject = false, PreserveObjectReferences = true};
+        var dcs1 = new DataContractSerializer(typeof(SerializationTestTypes.CustomClass), setting1);
+        var dcs2 = new DataContractSerializer(typeof(SerializationTestTypes.CustomClass), setting2);
+
+        MemoryStream ms = new MemoryStream();
+        dcs1.WriteObject(ms, dcrVariationsGoing);
+        ms.Position = 0;
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Do we want to verify the output written to `ms`?

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs (hunk 4 above)
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs]: I could've missed. Is this field used anywhere?

**Comment by shmao** [comment]
@v-yanbgu the new tests failed on CI. Can you please look into the failures? Thanks.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs (hunk 6 above)
Comment (v-yanbgu) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: As the method 'CompareRecursively' will verify the object type, properties, fields of the original object & deserialized object. I think we don't need to verify the output from ms. If you think we have the necessary, I will add the baseline to testing.

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs (hunk 4 above)
Comment (v-yanbgu) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs]: Sorry for this useless code. I'll remove it.

File: src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj (hunk 1 above)
Comment (v-yanbgu) [inline src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj]: We don't need it. Maybe it caused by some unknown factors from my solution, it was added. I will remove it.

**Comment by shmao** [comment]
`CompareRecursively` verifies that the object can make a round trip on the platform where the test runs. Verifying the output of the serialization, however, would help us make sure that the object can make round trips across platforms. For example, let's say we serialize `ObjectA` on Windows and Linux. If the test verified the de-serialized object only,  when the test passed on both platforms, we would know

On windows: ObjectA --> Payload A --> ObjectA
On Linux:      ObjectA --> Payload A' --> ObjectA

But `Payload A` can potentially be different with `Payload A'`. If that's the case, the cross platform serialization would fail.

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DCRTypeLibrary.cs [hunk 7]
```diff
@@ -27,4 +27,50 @@ public object Data2
...
+        public object unknownType1;
+
+        [DataMember]
+        public object unknownType2;
+    }
+
+    [DataContract]
+    public class CustomClass
+    {
+        [DataMember()]
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DCRTypeLibrary.cs]: Nit: the brackets are not necessary.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 8]
```diff
@@ -2860,6 +2860,103 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
         SerializationTestTypes.ComparisonHelper.CompareRecursively(value, actual);
     }
 
+    [Fact]
+    public static void DCS_BasicRoundtripDCRVariation1()
+    {
+        SerializationTestTypes.DCRVariations dcrVariationsGoing = new SerializationTestTypes.DCRVariations();
+        dcrVariationsGoing.unknownType1 = new SerializationTestTypes.Person();
+        dcrVariationsGoing.unknownType2 = new SerializationTestTypes.SimpleDC();
+        var setting1 = new DataContractSerializerSettings() { DataContractResolver = new SerializationTestTypes.SimpleResolver_Ser(), MaxItemsInObjectGraph = int.MaxValue, IgnoreExtensionDataObject = false, PreserveObjectReferences = true};
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Nit: put each assignment on one line.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 9]
```diff
@@ -2860,6 +2860,103 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
...
+        var dcr2 = new SerializationTestTypes.SimpleResolver_DeSer();                
+        var setting = new DataContractSerializerSettings() { MaxItemsInObjectGraph = int.MaxValue, IgnoreExtensionDataObject = false, PreserveObjectReferences = true };
+        var dcs = new DataContractSerializer(typeof(SerializationTestTypes.DCRVariations), setting);
+        string baseline = @"<DCRVariations z:Id=""1"" xmlns=""http://schemas.datacontract.org/2004/07/SerializationTestTypes"" xmlns:i=""http://www.w3.org/2001/XMLSchema-instance"" xmlns:z=""http://schemas.microsoft.com/2003/10/Serialization/""><unknownType1 z:Id=""2"" i:type=""a:SerializationTestTypes.Person"" xmlns:a=""System.Runtime.Serialization.Xml.Tests, Version=4.1.3.0, Culture=neutral, PublicKeyToken=9d77cc7ad39b68eb""><Age>0</Age><Name i:nil=""true""/></unknownType1><unknownType2 z:Id=""3"" i:type=""a:SerializationTestTypes.SimpleDC"" xmlns:a=""";
+
+        MemoryStream ms = new MemoryStream();
+        var xmlWriter = XmlDictionaryWriter.CreateTextWriter(ms);
+        dcs.WriteObject(xmlWriter, dcrVariationsGoing, dcr1);
+        CompareBaseline(dcrVariationsGoing, ms, baseline);
+        xmlWriter.Flush();
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: IMO it'd be better to call `Flush` before `CompareBaseline`

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 10]
```diff
@@ -2860,6 +2860,103 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
...
+        dcs.WriteObject(xmlWriter, dcrVariationsGoing);
+        CompareBaseline(dcrVariationsGoing, ms, baseline);
+        xmlWriter.Flush();
+        ms.Position = 0;
+        var xmlReader = XmlDictionaryReader.CreateTextReader(ms, XmlDictionaryReaderQuotas.Max);
+        var dcrVariationsReturning = dcs.ReadObject(xmlReader, false, dcr2);
+        SerializationTestTypes.ComparisonHelper.CompareRecursively(dcrVariationsGoing, dcrVariationsReturning);
+    }
+
+    public static void CompareBaseline(object t, MemoryStream ms,string baseline)
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: The method should be private. Can we change the method to `CompareBaseline(string baseline, MemoryStream outputStream)`? The object `t` seems not very helpful.

**Comment by shmao** [comment]
@v-yanbgu The tests are still failing on CI. Can you please take a look? Thanks!

**Comment by shmao** [comment]
@v-yanbgu The following tests failed on CI,

```
Test Result (4 failures / +4)
DataContractSerializerTests.DCS_BasicRoundtripDCRVariation2
DataContractSerializerTests.DCS_BasicRoundtripDCRVariation3
DataContractSerializerTests.DCS_BasicRoundtripDCRVariation4
DataContractSerializerTests.DCS_BasicRoundtripDCRVariation1

There were xunit test failures in System.Runtime.Serialization.Xml.ReflectionOnly.Tests
```

According to the [build log](https://ci.dot.net/job/dotnet_corefx/job/master/job/windows_nt_release_prtest/6697/testReport/junit/(root)/DataContractSerializerTests/DCS_BasicRoundtripDCRVariation2/), the root cause of the failure is that `SimpleResolver_Ser.TryResolveType` uses `dcType.Assembly.FullName` as type's namespace. The namespace became different when running in different assemblies. As the baseline string was for the test running in `System.Runtime.Serialization.Xml.Tests`, the test failed when running in ` System.Runtime.Serialization.Xml.ReflectionOnly.Tests`.

Do you know if there is any reason we wanted to use the assembly's full name as the namespace? I wondered if we can use a hard-coded namespace in the resolver.

**Comment by v-yanbgu** [comment]
Resolve 4 merge conflicts. So update codes.

**Comment by shmao** [comment]
@v-yanbgu can you please confirm if the tests passed when targeting `netfx`?

**Comment by v-yanbgu** [comment]
With below steps running tests on netfx which you provide in the email. 
When build targeting to netfx, the tests can pass.
1. From corefx root run: build.cmd -framework:netfx
2. msbuild <path-to-test-csproj> /t:rebuildandtest /p:targetgroup=netfx

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 11]
```diff
@@ -2858,6 +2858,131 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
...
+    public static void DCS_BasicRoundtripDCRVariation1()
+    {
+        SerializationTestTypes.DCRVariations dcrVariationsGoing = new SerializationTestTypes.DCRVariations();
+        dcrVariationsGoing.unknownType1 = new SerializationTestTypes.Person();
+        dcrVariationsGoing.unknownType2 = new SerializationTestTypes.SimpleDC();
+        var setting1 = new DataContractSerializerSettings()
+        {
+            DataContractResolver = new SerializationTestTypes.SimpleResolver_Ser(),
+            MaxItemsInObjectGraph = int.MaxValue,
+            IgnoreExtensionDataObject = false,
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Please keep necessary settings only.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 12]
```diff
@@ -2858,6 +2858,131 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
...
+            PreserveObjectReferences = true
+        };
+        var setting2 = new DataContractSerializerSettings()
+        { DataContractResolver = new SerializationTestTypes.SimpleResolver_DeSer(),
+            MaxItemsInObjectGraph = int.MaxValue,
+            IgnoreExtensionDataObject = false,
+            PreserveObjectReferences = true
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Please keep necessary settings only.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 13]
```diff
@@ -2858,6 +2858,131 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
...
+        dcrVariationsGoing.unknownType1 = new SerializationTestTypes.Person();
+        dcrVariationsGoing.unknownType2 = new SerializationTestTypes.SimpleDC();
+        var dcr1 = new SerializationTestTypes.SimpleResolver_Ser();
+        var dcr2 = new SerializationTestTypes.SimpleResolver_DeSer();
+        var setting = new DataContractSerializerSettings()
+        {
+            DataContractResolver = dcr2,
+            MaxItemsInObjectGraph = int.MaxValue,
+            IgnoreExtensionDataObject = false,
+            PreserveObjectReferences = true
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Please keep the necessary settings only.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 14]
```diff
@@ -2858,6 +2858,131 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
...
+    }
+
+    private static void CompareBaseline(string baseline, MemoryStream ms)
+    {
+        ms.Position = 0;
+        string actualOutput = new StreamReader(ms).ReadToEnd();
+        var result = Utils.Compare(baseline, actualOutput);
+        Assert.True(result.Equal, string.Format("{1}{0}Test failed.{0}Expected: {2}{0}Actual: {3}",
+                Environment.NewLine, result.ErrorMessage, baseline, actualOutput));
+    }
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Nit: a blank line after the brace.

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs [hunk 15]
```diff
@@ -168,4 +168,55 @@ public override Type ResolveName(string typeName, string typeNamespace, Type dec
             return t;
         }
     }
+
+    [Serializable]
+    public class SimpleResolver_Ser : DataContractResolver
+    {
+        public string defaultNS = "http://schemas.datacontract.org/2004/07/";
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs]: The field should be `private static readonly`.

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs [hunk 16]
```diff
@@ -168,4 +168,55 @@ public override Type ResolveName(string typeName, string typeNamespace, Type dec
...
+        public override Type ResolveName(string typeName, string typeNamespace, Type declaredType, DataContractResolver KTResolver)
+        {
+            throw new NotImplementedException("Deserialization is supposed to be handled by the SimpleResolver_DeSer resolver.");
+        }
+    }
+
+    [Serializable]
+    public class SimpleResolver_DeSer : DataContractResolver
+    {
+        public string defaultNS = "http://schemas.datacontract.org/2004/07/";
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/DataContractResolverLibrary.cs]: The field should be `private static readonly`. And the field name should be `s_defaultNs ` accroding to https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 17]
```diff
@@ -2858,6 +2858,131 @@ public static void DCS_BasicRoundTripResolveEnumStructTypes()
...
+        dcrVariationsGoing.unknownType2 = new SerializationTestTypes.SimpleDC();
+        var setting1 = new DataContractSerializerSettings()
+        {
+            DataContractResolver = new SerializationTestTypes.SimpleResolver_Ser(),
+            MaxItemsInObjectGraph = int.MaxValue,
+            IgnoreExtensionDataObject = false,
+            PreserveObjectReferences = true
+        };
+        var setting2 = new DataContractSerializerSettings()
+        { DataContractResolver = new SerializationTestTypes.SimpleResolver_DeSer(),
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: The code should start on a new line.

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/ObjRefSample.cs [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+{
+    #region Simple Class Structs with Ref
+
+    [DataContract(IsReference = true)]
+    public class SimpleDC
+    {
+        [DataMember]
+        public string Data;
+        public SimpleDC() { }
+        public SimpleDC(bool init)
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/ObjRefSample.cs]: Do we use this constructor anywhere?

File: src/System.Runtime.Serialization.Xml/tests/ReflectionOnly/System.Runtime.Serialization.Xml.ReflectionOnly.Tests.csproj [hunk 19]
```diff
@@ -24,14 +24,27 @@
...
+    </Compile>
+    <Compile Include="$(TestSourceFolder)..\SerializationTestTypes\DCRSampleType.cs">
+      <Link>SerializationTestTypes\DCRSampleType.cs</Link>
+    </Compile>
+    <Compile Include="$(TestSourceFolder)..\SerializationTestTypes\DCRTypeLibrary.cs">
+      <Link>SerializationTestTypes\DCRTypeLibrary.cs</Link>
+    </Compile>
+    <Compile Include="$(TestSourceFolder)..\SerializationTestTypes\Primitives.cs">
+      <Link>SerializationTestTypes\Primitives.cs</Link>
+    </Compile>
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/ReflectionOnly/System.Runtime.Serialization.Xml.ReflectionOnly.Tests.csproj]: These changes seems unnecessary. Can you please revert them?

File: src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/ObjRefSample.cs (hunk 18 above)
Comment (v-yanbgu) [inline src/System.Runtime.Serialization.Xml/tests/SerializationTestTypes/ObjRefSample.cs]: It was referenced by other DCS tests which are going to port into core. So I keep it there.

**Comment by v-yanbgu** [comment]
When build targeting to netfx, the tests can pass.

**Comment by shmao** [comment]
@v-yanbgu Thanks a lot for adding the tests!

**Pull request merged**